server/aesdsocket.c

```diff
@@ -11,26 +11,119 @@
 #include <arpa/inet.h>
 #include <sys/stat.h>
 #include <fcntl.h>
+#include <sys/queue.h>
+#include <pthread.h>
+#include <time.h>
+#include <sys/time.h>
 
 #define LOG_FILE_LOC "/var/tmp/aesdsocketdata"
 #define PORT "9000"
-#define MAX_BUFFER_SIZE 256
+#define MAX_BUFFER_SIZE 1024
+#define TIMESTAMP_INTERVAL 10 // seconds
+
+pthread_mutex_t aesdsock_mutex;
+pthread_mutex_t thread_list_mutex;
+pthread_mutex_t timer_mutex;
+pthread_t timer_thread;
+time_t last_timestamp;
+
+bool signal_exit= false;
+
+struct thread_data_s
+{
+    pthread_t id;
+    int cfd;
+    char peer_ip[INET6_ADDRSTRLEN];
+    bool thread_complete_flag;
+    SLIST_ENTRY(thread_data_s) next;
+};
 
 int sockfd = -1;
+SLIST_HEAD(thread_list, thread_data_s) thread_head = SLIST_HEAD_INITIALIZER(thread_head);
+
+// Function declarations
+void cleanup_and_exit(int status);
+void signal_handler(int signo);
+void setup_signal_handler();
+void *thread_function(void *arg);
+void handle_client_connection(int client_fd);
+void daemonize();
+void add_thread(int client_fd, const char *ip_addr);
+void setup_timer();
+void timer_handler(int signo);
 
 void cleanup_and_exit(int status)
 {
     syslog(LOG_INFO, "Closing aesdsocket application");
     closelog();
     close(sockfd);
     remove(LOG_FILE_LOC);
+
+    // Lock the mutex before deallocating memory
+    if (pthread_mutex_lock(&thread_list_mutex) != 0)
+    {
+        syslog(LOG_ERR, "Error locking thread list mutex");
+        exit(EXIT_FAILURE);
+    }
+
+    // Deallocate memory and join threads
+    struct thread_data_s *entry;
+    SLIST_FOREACH(entry, &thread_head, next)
+    {
+        if (pthread_join(entry->id, NULL) != 0)
+        {
+            syslog(LOG_ERR, "Error joining thread");
+            exit(EXIT_FAILURE);
+        }
+        free(entry);
```

Comment:
@akka2103 I don't think it's safe to free entry within the loop, I think you'd need to use FOREACH_SAFE for this.  See valgrind errors at https://github.com/cu-ecen-aeld/assignments-3-and-later-akka2103/pull/5/files#diff-85261e472d268426712db558b85101dca2eb5ab21de43bcc50c1450ee339e7ecR125

Reply:
I tried using FOREACH_SAFE, but the complier was throwing error, even when i include queue,h
That is I just went with this